### PR TITLE
Update Scheduler Docs

### PR DESCRIPTION
**Pull request by mburke5678**
From BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1596472
**1.** _Wrong statements or examples that I can tell:
Predicates [3] are boolean functions, they return true/false. The example is wrong, it does not make sense to have a weigh in a predicate.  Both ServiceAffinity and LabelsPresence Configurable Predicates examples are wrong.

Configurable priorities [4] examples are wrong.
They apply only to one label. So label must be in singular "label", not "labels"

Weight can be any positive value. It does not need to range between 0 and 10. Actually, in [5] you can see a default weight of 10000 in a function._

**2.** Do we want to add: 
_Service-Anti-Affinity based on custom labels is not recommended to achieve equal distribution of pods across multiple zones... it's highly recommended to use failure-domain.beta.kubernetes.io/region and failure-domain.beta.kubernetes.io/zone node label_ 
https://access.redhat.com/solutions/3432401

### CODE REVIEW

**Comment by mburke5678** [comment]
@aveshagarwal @ravisantoshgudimetla  I worked you on this topic some time back via https://bugzilla.redhat.com/show_bug.cgi?id=1596472 

Can you PTAL at admin_guide/scheduling/scheduler.adoc to address Jorge's concerns in  https://bugzilla.redhat.com/show_bug.cgi?id=1596472

**Comment by mburke5678** [comment]
@aveshagarwal @ravisantoshgudimetla PTAL at my changes and recommend any other changes?

Jorge Luis Tudela Gonzalez de Riancho
The scheduler doc topic [1] contains some bad examples and inaccurate statements.
...someone from Engineering should deeply review this. 

https://bugzilla.redhat.com/show_bug.cgi?id=1596472

**Comment by mburke5678** [comment]
@aveshagarwal @ravisantoshgudimetla PTAL at my changes and recommend any other changes?

Jorge Luis Tudela Gonzalez de Riancho
The scheduler doc topic [1] contains some bad examples and inaccurate statements.
...someone from Engineering should deeply review this. 

https://bugzilla.redhat.com/show_bug.cgi?id=1596472

**Comment by aveshagarwal** [comment]
@mburke5678 Could you remove weight from the following stanza: https://github.com/openshift/openshift-docs/pull/10742/files#diff-c881c052264120bfc3a506b42de92c15R481

```
         "name":"RackPreferred",
         "weight" : "1"
         "argument":{
            "labelsPresence":{
               "labels":[
                  "rack"
            "labelsPresence:"{
                "labels:"[
                - "region"
                presence: true
```

**Comment by aveshagarwal** [comment]
@mburke5678 also the above part is incorrectly formatted.

File: admin_guide/scheduling/scheduler.adoc [hunk 1]
```diff
@@ -634,7 +629,7 @@ concentration of pods.
    ]
 ----
 <1> Specify a name for the priority.
-<2> Specify a weight from 1 (bad fit) to 10 (best fit).
+<2> Specify a weight. Weight can be any positive value.
```
Comment (aveshagarwal) [inline admin_guide/scheduling/scheduler.adoc]: it could be negative too.

Comment (aveshagarwal) [reply admin_guide/scheduling/scheduler.adoc]: Non-zero.

File: admin_guide/scheduling/scheduler.adoc [hunk 2]
```diff
@@ -657,22 +652,20 @@ regardless of its value.
 "predicates":[
```
Comment (aveshagarwal) [inline admin_guide/scheduling/scheduler.adoc]: should be priorities.

File: admin_guide/scheduling/scheduler.adoc [hunk 3]
```diff
@@ -657,22 +652,20 @@ regardless of its value.
...
          "name":"<name>", <1>
-         "weight" : "1" <2>
-         "argument":{
+          "argument":{
             "labelsPresence":{
                "labels":[
-                  "<label>" <3>
-                presence: true/false <4>
+                  "<label>" <2>
+                presence: true/false <3>
```
Comment (aveshagarwal) [inline admin_guide/scheduling/scheduler.adoc]: the above part seems completely incorrect. We are talking about priorities and label preference but the above shows predicate and label presence. The above should be fixed.

Comment (mburke5678) [reply admin_guide/scheduling/scheduler.adoc]: @aveshagarwal I changed these parameters back to how they were in previous version, before the changes to the scheduling topic were made. 
Did you see  @jtudelag comments in the BZ: [_scheduling based on custom labels does not work... Most of the examples then do not apply...]( https://bugzilla.redhat.com/show_bug.cgi?id=1596472#c1). I am not sure what needs to be changed.

Comment (jtudelag) [reply admin_guide/scheduling/scheduler.adoc]: @mburke5678, I guess what @aveshagarwal means is that first of all, this is a priority function, not a predicate. Then it is "LabelPreference", not "LabelsPresence".

Also, as a function, It always should have a weight. It is still correct in OCP 3.5 examples:
https://docs.openshift.com/container-platform/3.5/admin_guide/scheduler.html#configurable-priority-functions

Comment (mburke5678) [reply admin_guide/scheduling/scheduler.adoc]: So, now I have Configurable Priority: `LabelPreference`, with `weight`, and `label` singlular/
Configurable Predicate: `LabelsPresence`, no weight and `labels` plural. Also has `presence: true/false`.
Need a definition for `LabelPreference`. They are slightly different
_LabelPreference prefers nodes that have a particular label defined or not, regardless of its value._
_LabelsPresence checks whether a particular node has a certain label defined or not, regardless of its value._

Comment (aveshagarwal) [reply admin_guide/scheduling/scheduler.adoc]: > @mburke5678, I guess what @aveshagarwal means is that first of all, this is a priority function, not a predicate. Then it is "LabelPreference", not "LabelsPresence".

> Also, as a function, It always should have a weight. It is still correct in OCP 3.5 examples:
https://docs.openshift.com/container-platform/3.5/admin_guide/scheduler.html#configurable-priority-functions

+1

> Main question is if this is still honoured by the scheduler in OCP. 

If I understand your questions correctly so yes LabelPreference and LabelPresence still work.

> Also, the explanation of what the LabelPreference function does should be re-written, as now it is just a copy paste of the LabelPresence predicate. @aveshagarwal should answer these questions, I dont have enought knowledge to do it ;(.

+1

Comment (mburke5678) [reply admin_guide/scheduling/scheduler.adoc]: @aveshagarwal 
"If true, higher priority is given to nodes that have the label"

What is "true" in this case? The presence of the specified label on a node?

`**_LabelPreference_** give priority based on the specified label. 
If the label is present on a node, that node is given priority.
If no label is specified, priority is given to nodes that do not have a label.`

Comment (aveshagarwal) [reply admin_guide/scheduling/scheduler.adoc]: > What is "true" in this case? The presence of the specified label on a node?

Yes.

**Comment by mburke5678** [comment]
@wjiangjay PTAL

**Review by jtudelag — APPROVED** [review]

File: admin_guide/scheduling/scheduler.adoc [hunk 4]
```diff
@@ -241,6 +241,12 @@ you require.
 # master-restart api master-restart controllers
 ----
 
+[NOTE]
+====
+There are cases where using configuring scheduler policy to use custom labels does not spread pod as expected.
```
[the remark this answers is not in the document]
Comment (aveshagarwal) [reply admin_guide/scheduling/scheduler.adoc]: Would be good to mention that is only about ServiceAntiAffinity specifically.

[the remark this answers is not in the document]
Comment (mburke5678) [reply admin_guide/scheduling/scheduler.adoc]: @jtudelag You are correct. I saw that, too; but hadn't pushed the change yet. Thank you. 
@aveshagarwal I reworded the note to mention ServiceAntiAffinity and moved it to that section.
https://github.com/openshift/openshift-docs/pull/10742/files#diff-c881c052264120bfc3a506b42de92c15R648

[the remark this answers is not in the document]
Comment (aveshagarwal) [reply admin_guide/scheduling/scheduler.adoc]: @mburke5678 lgtm.

[the remark this answers is not in the document]
Comment (jtudelag) [reply admin_guide/scheduling/scheduler.adoc]: @mburke5678 lgtm too ;).

File: admin_guide/scheduling/scheduler.adoc [hunk 5]
```diff
@@ -634,64 +627,54 @@ concentration of pods.
...
-            "labelsPresence":{
-               "labels":[
-                  "<label>" <3>
-                presence: true/false <4>
+         "weight" : "1"  <2>
+          "argument":{
+            "labelsPreference":{
+               "label":[
+                  "label":[  <3>
+                  "rack"
```
Comment (wjiangjay) [inline admin_guide/scheduling/scheduler.adoc]: Make another indent here.

File: admin_guide/scheduling/scheduler.adoc (hunk 3 above)
Comment (wjiangjay) [reply admin_guide/scheduling/scheduler.adoc]: presence should be quoted

File: admin_guide/scheduling/scheduler.adoc [hunk 6]
```diff
@@ -412,26 +412,23 @@ same label values as that node.
...
             }
          }
       }
    ],
 ----
 <1> Specify a name for the predicate.
-<2> Specify a weight from 1 (bad fit) to 10 (best fit).
-<3> Specify a label for matching.
+<2> Specify a label for matching.
 For example:
```
Comment (wjiangjay) [inline admin_guide/scheduling/scheduler.adoc]: `For example:` should be started with newline

**Comment by wjiangjay** [comment]
Others LGTM

**Comment by mburke5678** [comment]
@avesh 
Can you take a quick look at the changes I made from @wjiangjay comments above?
Line 659  presence: true/false <3>  Presence should have be quoted
SB
`presence: "true/false"`

Line 667 Two labels here?
Should be:
```
"labelsPreference":{
     "labels":[
           "<label>"
  ]
```

**Comment by mburke5678** [comment]
@openshift/team-documentation PTAL

File: admin_guide/scheduling/scheduler.adoc [hunk 7]
```diff
@@ -412,26 +412,24 @@ same label values as that node.
...
                ]
             }
          }
       }
    ],
 ----
 <1> Specify a name for the predicate.
-<2> Specify a weight from 1 (bad fit) to 10 (best fit).
-<3> Specify a label for matching.
+<2> Specify a label for matching.
```
Comment (kalexand-rh) [inline admin_guide/scheduling/scheduler.adoc]: s/for matching/to match ?

File: admin_guide/scheduling/scheduler.adoc [hunk 8]
```diff
@@ -446,31 +444,28 @@ see xref:../../admin_guide/scheduling/pod_placement.adoc#controlling-pod-placeme
 Multiple-level labels are also supported. Users can also specify all pods for a service to
 be scheduled on nodes within the same region and within the same zone (under the region).
 
-**_LabelsPresence_** checks whether a particular node has a certain label
-defined or not, regardless of its value. Matching by label can be useful, for
-example, where nodes have their physical location or status defined by labels.
+**_LabelsPresence_** checks whether a particular node has a specific label. The labels create node _groups_ that the
+`LabelPreference` priority uses. Matching by label can be useful, for example, where nodes have their physical location or status defined by labels.
```
Comment (kalexand-rh) [inline admin_guide/scheduling/scheduler.adoc]: s/\`LabelPreference\`/\`labelPreference\` ?

File: admin_guide/scheduling/scheduler.adoc [hunk 9]
```diff
@@ -446,31 +444,28 @@ see xref:../../admin_guide/scheduling/pod_placement.adoc#controlling-pod-placeme
 Multiple-level labels are also supported. Users can also specify all pods for a service to
 be scheduled on nodes within the same region and within the same zone (under the region).
 
-**_LabelsPresence_** checks whether a particular node has a certain label
-defined or not, regardless of its value. Matching by label can be useful, for
-example, where nodes have their physical location or status defined by labels.
+**_LabelsPresence_** checks whether a particular node has a specific label. The labels create node _groups_ that the
```
Comment (kalexand-rh) [inline admin_guide/scheduling/scheduler.adoc]: s/\**_LabelsPresence\_**\The \`labelsPresence\` parameter

File: admin_guide/scheduling/scheduler.adoc [hunk 10]
```diff
@@ -446,31 +444,28 @@ see xref:../../admin_guide/scheduling/pod_placement.adoc#controlling-pod-placeme
...
             }
          }
       }
    ],
 ----
 <1> Specify a name for the predicate.
-<2> Specify a weight from 1 (bad fit) to 10 (best fit).
-<3> Specify a label for matching.
-<4> Specify whether the labels are required.
+<2> Specify a label for matching.
```
Comment (kalexand-rh) [inline admin_guide/scheduling/scheduler.adoc]: s/for matching/to match

File: admin_guide/scheduling/scheduler.adoc [hunk 11]
```diff
@@ -446,31 +444,28 @@ see xref:../../admin_guide/scheduling/pod_placement.adoc#controlling-pod-placeme
...
       {
          "name":"<name>", <1>
-         "weight" : "1" <2>
          "argument":{
             "labelsPresence":{
                "labels":[
-                  "<label>" <3>
-                presence: true/false <4>
+                  "<label>" <2>
+                presence: "true/false" <3>
```
Comment (kalexand-rh) [inline admin_guide/scheduling/scheduler.adoc]: I'd s/"true/false"/"true" and specify the valid parameters in annotation 3.

File: admin_guide/scheduling/scheduler.adoc [hunk 12]
```diff
@@ -634,64 +628,53 @@ concentration of pods.
    ]
 ----
 <1> Specify a name for the priority.
-<2> Specify a weight from 1 (bad fit) to 10 (best fit).
+<2> Specify a weight. Weight can be any non-zero positive value.
```
Comment (kalexand-rh) [inline admin_guide/scheduling/scheduler.adoc]: Is there no longer a range for valid weights?
s/Weight can be/Enter

File: admin_guide/scheduling/scheduler.adoc [hunk 13]
```diff
@@ -634,64 +628,53 @@ concentration of pods.
...
          "argument":{
             "serviceAffinity":{
-               "labels":[
-                  "rack"
+               "label":[    <3>
+                  "rack"   
 ----
+<1> Specify a name for the priority.
+<2> Specify a weight. Weight can be any non-zero positive value.
+<3> Specify a label for matching.
```
Comment (kalexand-rh) [inline admin_guide/scheduling/scheduler.adoc]: s/for matching/to match

File: admin_guide/scheduling/scheduler.adoc [hunk 14]
```diff
@@ -634,64 +628,53 @@ concentration of pods.
...
 
-**_LabelPreference_** prefers nodes that have a particular label defined,
-regardless of its value.
+[NOTE]
+====
+There are situations when using `ServiceAntiAffinity` based on custom labels does not spread pod as expected.
```
Comment (kalexand-rh) [inline admin_guide/scheduling/scheduler.adoc]: s/There are situations when/In some situations,

File: admin_guide/scheduling/scheduler.adoc [hunk 15]
```diff
@@ -634,64 +628,53 @@ concentration of pods.
...
 
-**_LabelPreference_** prefers nodes that have a particular label defined,
-regardless of its value.
+[NOTE]
+====
+There are situations when using `ServiceAntiAffinity` based on custom labels does not spread pod as expected. 
+See link:https://access.redhat.com/solutions/3432401[this Red Hat Solution]. 
+====
+
+**_LabelPreference_** gives priority based on the specified label.
```
Comment (kalexand-rh) [inline admin_guide/scheduling/scheduler.adoc]: s/\**_LabelPreference\_**\The \`labelPreference\` parameter

File: admin_guide/scheduling/scheduler.adoc [hunk 16]
```diff
@@ -634,64 +628,53 @@ concentration of pods.
...
+         "weight" : "1"     <2>
          "argument":{
             "serviceAffinity":{
-               "labels":[
-                  "rack"
+               "label":[    <3>
+                  "rack"   
 ----
+<1> Specify a name for the priority.
+<2> Specify a weight. Weight can be any non-zero positive value.
```
Comment (kalexand-rh) [inline admin_guide/scheduling/scheduler.adoc]: s/Weight can be/Enter

File: admin_guide/scheduling/scheduler.adoc [hunk 17]
```diff
@@ -634,64 +628,53 @@ concentration of pods.
...
+                  "<label>"  <3>
                ]
             }
          }
       }
    ],
 ----
 <1> Specify a name for the priority.
-<2> Specify a weight from 1 (bad fit) to 10 (best fit).
+<2> Specify a weight. Weight can be any non-zero positive value.
```
Comment (kalexand-rh) [inline admin_guide/scheduling/scheduler.adoc]: s/Weight can be/Enter

**Review by kalexand-rh — COMMENTED** [review]
Couple of picks, suggestions, and questions.

**Comment by aveshagarwal** [comment]
@mburke5678 lgtm.

File: admin_guide/scheduling/scheduler.adoc (hunk 12 above)
Comment (mburke5678) [reply admin_guide/scheduling/scheduler.adoc]: @kalexand-rh No. This was incorrect.

File: admin_guide/scheduling/scheduler.adoc [hunk 18]
```diff
@@ -446,31 +444,28 @@ see xref:../../admin_guide/scheduling/pod_placement.adoc#controlling-pod-placeme
 Multiple-level labels are also supported. Users can also specify all pods for a service to
 be scheduled on nodes within the same region and within the same zone (under the region).
 
-**_LabelsPresence_** checks whether a particular node has a certain label
-defined or not, regardless of its value. Matching by label can be useful, for
-example, where nodes have their physical location or status defined by labels.
+The `abelsPresence` parameter checks whether a particular node has a specific label. The labels create node _groups_ that the
```
Comment (aveshagarwal) [inline admin_guide/scheduling/scheduler.adoc]: typo: abelsPresence (`L`) is missing

File: admin_guide/scheduling/scheduler.adoc [hunk 19]
```diff
@@ -446,44 +444,40 @@ see xref:../../admin_guide/scheduling/pod_placement.adoc#controlling-pod-placeme
...
       {
          "name":"<name>", <1>
-         "weight" : "1" <2>
          "argument":{
             "labelsPresence":{
                "labels":[
-                  "<label>" <3>
-                presence: true/false <4>
+                  "<label>" <2>
+                presence: "true" <3>
```
Comment (wjiangjay) [inline admin_guide/scheduling/scheduler.adoc]: here, the structure should be:

```
"predicates":[
      {
         "name":"<name>", (1)
         "argument":{
            "labelsPresence":{
               "labels":[
                  "<label>" (2)
                ],
                "presence": true (3)
            }
         }
      }
   ],
```

Comment (mburke5678) [reply admin_guide/scheduling/scheduler.adoc]: Made this change in a follow-up PR: https://github.com/openshift/openshift-docs/pull/11335

**Pull request merged**